### ISSUE TITLE
Nunjucks with multiple directories(for layouts, templates, partials) support.

### ISSUE THREAD

**Issue by hiteshbalraw** [mozilla/nunjucks#495]
Hi,

I am doing following.

    /*
    config._views is my views directory
	  ~/nunjucks-demo/views
             --views
                 -templates
                 -layouts
                 -partials

     engines => consolidate
     module => "nunjucks"
    */
    

    // express cofiguration
    app.engine(ext, engines[module]);
    app.set('view engine', ext);
    app.set('views', path.join(config._views, 'templates'));
    
    // nunjucks configuration
    var nunjucks = require('nunjucks');
    var env = nunjucks.configure(config._views, {
                autoescape: false,
                express: app
            });

    // nunjucks-demo/views/templates/index.html
    {% extends "layout/default.html" %}
    {% block content %}
        <section>
            <h1>{{entry.title}}</h1>
            <ul>
                <li><b>markdown:</b> {{ entry.markdown }}</li>
            </ul>
        </section>
    {% endblock %}
    
    // nunjucks-demo/views/layout/default.html
    <!DOCTYPE HTML>
    <html>
        <head>
            <title>{{entry.title}}</title>
            <link type="text/css" rel="stylesheet" href="/static/css/style.css">
            <script type="text/javascript" src="/static/js/script.js"></script>
        </head>
        <body>
            {% include "partials/header.html" %}
            {% block content %}{% endblock %}
            {% include "partials/footer.html" %}
        </body>
    </html>

But it doesn't work I have to make all the templates and layouts in single directory.

Layouts are not even extending. This same thing works in swig.

**Comment by ricordisamoa** [comment]
> {% extensds "layout/default.html" %}

typo?

**Comment by hiteshbalraw** [comment]
Yes it was typo. Now modified it.

**Comment by jlongster** [comment]
You can pass multiple directories to `nunjucks.configure` to load from. So: `nunjucks.configure(['views', 'views/templates', {})`.

Your question is not entirely clear. If you could explain the problem and show the error you're getting that would help.

**Comment by hiteshbalraw** [comment]
Hi,

What I mean to say was can we just configure like swig.
http://paularmstrong.github.io/swig/docs/#express

By specifing the single directory in express views where the template engine will look for the includes and extends call and main template directory is specified in nunjucks configuration.

**Comment by mcornella** [comment]
I think what @hiteshbalraw means is to be able to pass a `partials` option that will set the root folder of where to look for the nunjucks templates.
For instance, if we have the following structure:
```
views
├── layouts
├── partials
└── templates
```
and we pass `{ partials: 'views' }` in the context of `render` and `renderString`, then we don't need to specify the whole path to include/extend a file, but only the relative path to the root folder (in this case, `views`). This is, essentially, an alternative to having to use `nunjucks.configure(path)`.

This is used a lot by [consolidate.js](https://github.com/tj/consolidate.js), and I need this one specifically to use it with the [`metalsmith-in-place` plugin](https://github.com/superwolff/metalsmith-in-place#partials). If you need more context, [here's how consolidate.js calls the nunjucks renderString method](https://github.com/tj/consolidate.js/blob/master/lib/consolidate.js#L1014-L1025).

**Comment by carljm** [comment]
I'm afraid that doesn't clarify anything for me. I don't understand what advantage this complex system brings over just having a single root path for templates, passing it to `nunjucks.configure` (or passing it to a loader and then passing that loader in the opts), and referencing all template paths relative to that single root. Why would you want to split up your templates and have implicit behavior regarding which templates are looked up where?

And even if you do want that, why not just pass a list of paths with the various template roots you want, as jlongster showed above?

I don't understand "then we don't need to specify the whole path... but only the relative path." What is the "whole path"? You always specify a "relative path" when including a template in nunjucks; relative to the root template path you configured your environment/loader with. Why do you need different paths to be relative to different roots?

I don't understand why one would _want_ "an alternative to using `nunjucks.configure(path)`. You're going to have to tell your template engine where your templates live somehow; what exactly is wrong with that method?

The link to `consolidate.js` doesn't provide useful context for this request as far as I can see. I don't see any reference there to multiple different template paths.

I apologize if I'm being dense. Maybe someone needs to try a different angle on describing this, because at the moment I'm afraid I don't really have a clue what is being requested here or why it would be useful.

**Comment by carljm** [comment]
(It seems to me that perhaps express - and consolidate? - have a very different way of thinking about templating from how nunjucks - and other templating engines I've used - think about it. There are several other open bugs on nunjucks related to express integration that I have no idea how to approach. I'm increasingly inclined to remove all explicit express integration from nunjucks, and ask that people who are using the two together provide whatever integration layer makes the most sense as a separate project.)

**Comment by mcornella** [comment]
I'm sorry, I could have worded that better. It's a ridiculous request, actually. [Here's what the `partials` option would do in my setup:](https://github.com/superwolff/metalsmith-in-place#partials)

> ```
> {
>   ... 
>    partials: 'partials'
>   ... 
> }
> ```
> would mean that a partial at `partials/nav.html` can be used as `{{> nav }}`, and `partials/nested/footer.html` can be used as `{{> nested/footer }}`.

They're using handlebars, but it would be similar to an include with nunjucks. 
In that example, the `partials/` path doesn't need to be specified, because the template engine knows that's the base folder of the template file because the `consolidate.js` plugin _told them_ by passing the `partials` option to the templating engine.

Now, I thought this `partials` argument was some sort of standard in templating engines, but it seems that it isn't that common. So I do think now that there should be some middleware code in `consolidate.js` that transforms their `partials` parameter into whatever the specific templating engine is expecting, which in this case is a `nunjucks.configure` call with the template path. After all, the `partials` parameter is part of `consolidate.js` API, not yours.

**TL;DR** I'll submit this question over to `consolidate.js`, which is the middleware that does the integration in this case. Sorry for bothering!

**Comment by carljm** [comment]
No problem, thanks for the additional clarification! I appreciate you taking the time to educate me, I think I understand it now :-) And I agree that that feels too purpose-specific to belong in nunjucks proper.

**Comment by mcornella** [comment]
My pleasure! 

BTW, I'm just starting to use nunjucks but when I feel comfortable enough I'll take a look at one of those pesky bugs of yours. Stay strong! :smile:

**Comment by hiteshbalraw** [comment]
Hey Guys,

Thanks for the clarification.
